File: backend-api/pkg/authz/claims.go
package authz

import (
	"context"
)

type ctxKeyAuthClaims int

const (
	PrincipalClaimsKey ctxKeyAuthClaims = 1
)

// PrincipalClaims encapsulates the authenticated identity, tenant context, and resolved permissions.
type PrincipalClaims struct {
	IsAuthenticated      bool   `json:"isAuthenticated"`
	SessionID            string `json:"sessionId"`
	IdentityID           string `json:"identityId"`
	Email                string `json:"email"`
	FullName             string `json:"fullName"`
	ActiveOrganizationID string `json:"activeOrganizationId,omitempty"`
	OrganizationName     string `json:"organizationName,omitempty"`
	OrganizationSlug     string `json:"organizationSlug,omitempty"`
	OrganizationLogo     string `json:"organizationLogo,omitempty"`
	OrganizationKind     string `json:"organizationKind,omitempty"`
	OrganizationRole     string `json:"organizationRole,omitempty"`
}

// ExtractClaimsFromContext retrieves the PrincipalClaims stored in the context.
func ExtractClaimsFromContext(ctx context.Context) PrincipalClaims {
	if ctx != nil {
		if claims, ok := ctx.Value(PrincipalClaimsKey).(*PrincipalClaims); ok && claims != nil {
			return *claims
		}
	}

	return PrincipalClaims{IsAuthenticated: false}
}

// AddClaimsToContext stores the PrincipalClaims in the provided context.
func AddClaimsToContext(ctx context.Context, claims *PrincipalClaims) context.Context {
	return context.WithValue(ctx, PrincipalClaimsKey, claims)
}

File: frontend/src/lib/models/claims.ts
export interface PrincipalClaims {
	isAuthenticated: boolean;
	token: string;
	identityId: string;
	fullName: string;
	Email: string;
	activeOrganizationId: string;
	organizationSlug: string;
	organizationName: string;
	organizationKind: string;
	organizationRole: string;
}

export const ANONYMOUS_CLAIMS: PrincipalClaims = {
	isAuthenticated: false,
	token: '',
	identityId: '',
	fullName: '',
	Email: '',
	activeOrganizationId: '',
	organizationSlug: '',
	organizationName: '',
	organizationKind: '',
	organizationRole: ''
};

export class ClaimsService {
	private readonly claims: PrincipalClaims;

	constructor(claims: PrincipalClaims) {
		this.claims = claims;
	}

	public getClaims(): PrincipalClaims {
		return this.claims;
	}

	public isAuthenticated(): boolean {
		return this.claims.isAuthenticated;
	}

	public isManagementAdmin(): boolean {
		return (
			this.claims.isAuthenticated &&
			this.claims.organizationKind === 'management' &&
			this.claims.organizationRole === 'admin'
		);
	}
}
